Reduced rebuild times by reimplementing `GenCMessages` and fixing `vizInterface` dependencies in CMake

dpad opened the pull request:
* **Tickets addressed:** n/a
* **Review:** By file  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description

Currently, each time that Basilisk is rebuilt (`python conanfile.py`...), several files get re-generated even when there are no underlying changes. This causes a chain of unnecessary recompilations -- on my machine, a rebuild with no changes takes ~210 seconds.

This pull request fixes up how the generated file dependencies are tracked and gives this information directly to CMake. On my machine, a rebuild with no changes now takes ~3.5 seconds (it still rebuilds the auto draft modules, because I didn't bother to extract this yet). Changing a single message `Payload.h` file now builds in ~15 seconds.

The changes are as follows:

- **Reimplemented `GenCMessages.py` as a function directly inside CMake.**
  - Previously, C message interfaces were re-generated on every call to Conan, and had to be recompiled even if they hadn't changed. CMake believed these files were newer than their dependents, triggering a huge number of other compilations.
  - Reduced the scope of the generation. Some of the outputs of the `GenCMessages.py` script were no longer used or relevant, so the script was unnecessarily complicated. (The `.i` files were only necessary until commits 44cb3d538f7b321bf65408862ad2f83905c5040a and 40c81186c1b83323a1ddd913cc4dfff022b82062. Also, C++ message payloads didn't need to go through this script at all anymore.)
- **Fixed the dependencies for building `vizInterface` so that it recompiles only when they change.**
  - Previously, the protobuf message was being rebuilt on every run, triggering a recompilation. 

## Alternatives

For `GenCMessages.py`, an alternative is to rewrite the script in Python and call it directly from CMake instead of Conan. The main benefit of this would be that we have all the power of Python available in case we need to make the templates more complex. However, the downsides are that we need to be more manual on the CMake side, such as explicitly specifying the actual files that we expect the script to generate (CMake can't really track generated files dynamically without work-arounds).

For these reasons, and given that the templates have remained quite simple until now, I decided to reimplement `GenCMessages.py` in CMake directly.

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? If you didn't 
add or update any tests justify this choice. -->

- Built and ran all tests (including scenario tests) -- everything passed in all cases.
- Made some modifications, such as changing message files, adding and removing files, etc. and confirmed that correct targets were built as expected.

It would be very helpful for the maintainers to double-check my assumptions and to build and test this on various platforms, as I only tested on a single Linux (Ubuntu 22.04) instance. Also, please confirm that the `messaging.auto.i`, `messaging.header.auto.i`, and `cMsgCInterfacePy.auto.i` files are no longer used or necessary, as I believe that is the case and have removed them entirely.

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and 
completeness? -->

N/a

## Future work
<!-- What next steps can we anticipate from here, if any? -->

N/a

schaubh commented on requirements.txt:
```diff
@@ -1,6 +1,5 @@
 conan>=1.40.1, <2.00.0
 pandas
-parse>=1.18.0
```
I can confirm that `parse` is no longer needed as a required package.  However, can you please split out this commit into a separate commit?  Right now this change is not related to the rest of the commit.

schaubh commented on conanfile.py:
```diff
@@ -237,21 +237,7 @@ def imports(self):
             self.keep_imports = True
             self.copy("*.dll", "../Basilisk", "bin")
 
-    def generateMessageModules(self, originalWorkingDirectory):
```
I think this file (http://hanspeterschaub.info/basilisk/Learn/makingModules/makingModules-2.html) should be updated to explain to the user when they need to do a clean build again to catch changes in the message edits or message deletions?

dpad replied:
The behaviour you described is currently the behaviour expected on other (non-`messaging`) packages, e.g. `ExternalModules`. That is, the build system will keep any previously built but renamed/deleted modules, and it is up to the developer to manually clean the build (e.g. with `--clean`). For example, building with `--pathToExternalModules external` will create `Basilisk.ExternalModules`, but if you then rebuild without this flag, that module will still exist unless the user runs a `--clean` build.

However, I have added an additional commit to maintain the desired auto-generation behaviour for the `messaging` package specifically. Now, auto-generated message files will get created in `dist3/messaging`, but only the *currently existing* messages will be copied from `dist3/messaging` -> `dist3/Basilisk/architecture/messaging`. This way,
- Previously built messages still exist in `dist3/messaging`, so they don't need to be rebuilt unless their dependencies change.
- Only messages which have an existing `*Payload.h` file will actually get copied into the final Basilisk package, so deleted messages won't be importable (although their compiled versions still exist in `dist3/messaging`).

This behaviour complicates things a bit, because the default behaviour of CMake is to leave previously built files alone until the user runs a `clean`, but I think the implementation is still okay. Please have a look and let me know if this is acceptable or not.

schaubh replied:
you make a good point about the use of ``--clean``.  Maybe leave the original behavior, and we can just update the BSK documentation to explain when to use a `--clean` command?  

I don't mind the build behavior changing to be more consistent with how other elements are built (i.e. the modules folder, etc.). I just want to be explicit to the develop on what behavior to expect. :-)

dpad replied:
I've reverted the "Clean up stale message files..." change and went back to requiring `--clean`. I've added a caution to the documentation page you mentioned.

Note: I left the revert as an explicit revert commit in case we ever want to refer back to the implementation that only exports messages that actually exist. This might be useful in some long-term future build system changes (for reference, I'm considering possible ways to completely separate building of `ExternalModules` from building all of Basilisk).

schaubh reviewed (changes requested):
Interesting branch.  I text on macOS with latest Xcode and OS. I can confirm that builds in Xcode, or using make file, are cut in half if there are no changes to the code.  

In addition to the questions posted in this review, please also add:
- update BSK release notes to briefly discuss this change

I didn't see any issues with this branch, but I asked @spiggottCO to take a look at this as well as he helped write the prior version of this C message wrapper file auto-coding process.

schaubh commented:
The CI builds are failing here because the GTest suite of unit tests are not running?  Looking at the changes, I'm not sure what is happening here.  The errors are consistent across the Linux and Windows CI builds.

dpad commented:
Had a quick look at the test, that is indeed utterly bizarre. I'll have a deeper look after the weekend.

schaubh commented:
On my macOS laptop I was able to build this branch and run all the tests, including the Google test suite.

dpad commented:
@schaubh I cannot reproduce on my Linux machine. The same error is occurring on other branches, so it's not related to these changes.

I did a quick investigation but couldn't find any obvious reason. For reference, it seems that the CI runs a fairly old version of CMake (3.16.3), but I'm still able to run the `ctest` command successfully on that version on my machine.

schaubh commented:
@dpad , the issue was with how the `ctest` command was being called to create an optional output.  I found a fix on another BSK fork and created a PR to push this shortly to main BSK.  We can then rebase your branch and retest.  This CI issue should then go away.

schaubh commented:
Mm, this branch was building before on my macOS system now it isn't anymore.  The CI builds are both failing for Linux and Windows.  I'm attaching my console log.  The issues looks related to the message element.
[bskBuildPy311.txt](https://github.com/user-attachments/files/16474512/bskBuildPy311.txt)

dpad commented:
@schaubh Thanks for the logs! Yeah sorry, that's my mistake, I was having the opposite problem of what you experienced here, where the symbolic link wouldn't work because it already existed. I wonder if there's a difference between Python versions. In any case, I should have made a more robust check for it.

dpad commented:
@schaubh I've added an explicit check for the symbolic link and reverted back to having to run `--clean` (also added documentation about this) as per your request. Please have a look and let me know if you run into any other issues.

schaubh commented:
> @schaubh I've added an explicit check for the symbolic link and reverted back to having to run `--clean` (also added documentation about this) as per your request. Please have a look and let me know if you run into any other issues.

Thanks for the update.  The CI build completed without issues and it built again on my macOS laptop.  I'm going to chat later today with the original author of this code to get his feedback.

schaubh commented on docs/source/Learn/makingModules/makingModules-2.rst:
```diff
@@ -37,6 +37,8 @@ To create message objects in C and C++ modules, you need to run ``python3 conanf
 
 Running ``python3 conanfile.py`` re-creates the IDE project file that will now include access to the new message definition.
 
+.. caution::
```
Minor thing, could you please reduce the length of the first line of the commit message here?  It is too long for
GitHub to render properly.  Otherwise documentation built well.

dpad replied:
Done and rebased as well.

schaubh reviewed (commented):
Just a minor comment right now on this PR.  I plan to get more feedback from a collaborator who worked on the original code.

schaubh commented:
Thanks for doing the rebase and small change.  Had a great discussion with @spiggottCO yesterday afternoon about this branch who will take a look as well.  He most of the original code and was excited by your approach.  I don't have further comments at this time, I'll give @spiggottCO a few days to look over this branch and comment. Classes start up here in 2.5 weeks, and next week is an astro conference where many are presenting.  Hectic times, but I'm excited to incorporated these improvements you are making.

spiggottCO commented:
This is a great change.  Works for me and it should greatly reduce building pain across the project.

spiggottCO commented on src/architecture/messaging/cMsgCInterface/msg_C.h.in:
```diff
@@ -0,0 +1,55 @@
+/* @MSG_AUTOSOURCE_LICENSE@ */
```
File isn't actually new, but git thinks it is for some reason.

schaubh replied:
Odd, I see what you mean.  But, this should be good to go though.

spiggottCO reviewed (approved):
Looks great as-is.

schaubh reviewed (changes requested):
Howdy, this branch is essentially good to go.  Could you please rebased on latest develop and resolve small conflicts.  The new build system you helped create is now on `develop`.

dpad commented:
@schaubh Rebase done. Seemingly still working as expected for me, but would be good to reconfirm on your end.

schaubh commented:
@dpad , thanks for the support.  I'll run some local checks on this as well and merge if no issues are found.

Are you still looking at making a branch providing conan 2.x support?  You had this support on your original PR.  Myself, I'm pondering how to split up the BSK data files and examples from the wheels to reduce the wheel size.  I have some ideas to play with.

schaubh reviewed (approved):
I build with all modules, including external folder modules, and everything build and ran fine.  Thanks for this contribution.

dpad commented:
> Myself, I'm pondering how to split up the BSK data files and examples from the wheels to reduce the wheel size. I have some ideas to play with.

I have a suspicion as well that the way we're building some of the modules is greatly increasing the size. For example, I think each message module comes with its own copy of `SysModel`, etc. I suspect that if we aggregated all the individual modules together into single modules, instead of individual ones (e.g. all the `simulation/_blahblah.so` into a single `simulation.so`), we might significantly reduce the file size as well. But I'm not super versed in SWIG and this kind of development yet so not sure how much is possible.
